### PR TITLE
Revamp parachain-selection part 2

### PR DESCRIPTION
**Pull request by kianenigma**
- [x] Make tests work again
- [x] TODOs in the code.
fixes parts of https://github.com/paritytech/statemint/issues/9
related to https://github.com/paritytech/statemint/issues/23

### CODE REVIEW

File: node/src/chain_spec.rs [hunk 1]
```diff
@@ -126,12 +125,14 @@ fn statemint_testnet_genesis(
 		},
 		pallet_sudo: statemint_runtime::SudoConfig { key: root_key.clone() },
 		parachain_info: statemint_runtime::ParachainInfoConfig { parachain_id: id },
-		pallet_collator_selection: statemint_runtime::CollatorSelectionConfig { invulnerables: vec![root_key] },
+		pallet_collator_selection: statemint_runtime::CollatorSelectionConfig {
+			invulnerables: vec![root_key],
```
Comment (kianenigma) [inline node/src/chain_spec.rs:129]: this doesn't work yet, need to use `get_pair_from_seed` directly.

File: pallets/collator-selection/src/lib.rs [hunk 2]
```diff
@@ -115,21 +131,30 @@ pub mod pallet {
...
 		/// The weight information of this pallet.
 		type WeightInfo: WeightInfo;
 	}
 
 	/// Basic information about a collation candidate.
 	#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug)]
-	pub struct CandidateInfo<AccountId, Balance, BlockNumber> {
-		/// Account ID.
+	pub struct CandidateInfo<AccountId, Balance, BlockNumber, Keys> {
+		/// Account id
```
Comment (kianenigma) [inline pallets/collator-selection/src/lib.rs:153]:
```suggestion
		/// Account identifier. 
```

File: pallets/collator-selection/src/lib.rs [hunk 3]
```diff
@@ -221,21 +258,28 @@ pub mod pallet {
...
+		fn on_initialize(n: T::BlockNumber) -> Weight {
+			if (n % T::SessionLength::get().max(One::one())).is_zero() {
+				let mut collators = Self::invulnerables();
+				collators.extend(
+					Self::candidates().into_iter().map(|c| (c.who, c.keys)).collect::<Vec<_>>(),
+				);
+				Self::deposit_event(Event::EpochChanged(
+					collators.iter().cloned().map(|(c, _)| c).collect::<Vec<_>>(),
+				));
+				T::SessionHandlers::on_new_session(true, &collators, &collators);
```
Comment (kianenigma) [inline pallets/collator-selection/src/lib.rs:276]: for now I am blindly passing `changed == true`, could be made better.

**Comment by kianenigma** [comment]
@apopiak @riusricardo do you guys spawn test networks for testing things like #41? As soon as I manage to get this to build, we can also include this PR and test the process of generating keys and submitting them.

**Comment by apopiak** [comment]
Yeah we spawned a test network with `polkadot-launch` and played around with it.

File: pallets/collator-selection/src/lib.rs [hunk 4]
```diff
@@ -196,7 +196,9 @@ pub mod pallet {
 	#[pallet::genesis_build]
 	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
 		fn build(&self) {
-			// TODO: prevent duplicate.
+
+			let duplicate_invulnerables = self.invulnerables.iter().map(|x| x).cloned().collect::<std::collections::BTreeSet<_>>();
+			assert!(duplicate_invulnerables.len()  == self.invulnerables.len(), "duplicate invulnerables in genesis.");
```
Comment (kianenigma) [inline pallets/collator-selection/src/lib.rs:201]: I don't quite get this; how is this preventing dupes? 

If incorrect, you can `clone`, `dedup`, and `compare`. Alternatively, keep sorted and binary search each element.

Comment (kianenigma) [reply pallets/collator-selection/src/lib.rs:201]: Ahh I am blind, the `BtreeSet`. 

The `.map` is redundant though, otherwise okay.

**Review by JesseAbram — APPROVED** [review]

File: node/src/chain_spec.rs [hunk 5]
```diff
@@ -13,26 +13,44 @@
...
+use sp_consensus_aura::sr25519::AuthorityId as AuraId;
 use hex_literal::hex;
-use polkadot_service::ParaId;
+
 /// Specialized `ChainSpec` for the normal parachain runtime.
 pub type ChainSpec = sc_service::GenericChainSpec<statemint_runtime::GenesisConfig, Extensions>;
 pub type StatemineChainSpec = sc_service::GenericChainSpec<statemine_runtime::GenesisConfig, Extensions>;
 
+const STATEMINT_ED: Balance = statemine_runtime::EXISTENTIAL_DEPOSIT;
+const STATEMINE_ED: Balance = statemine_runtime::EXISTENTIAL_DEPOSIT;
```
Comment (apopiak) [inline node/src/chain_spec.rs:31]:
```suggestion
const STATEMINT_ED: Balance = statemint_runtime::EXISTENTIAL_DEPOSIT;
const STATEMINE_ED: Balance = statemine_runtime::EXISTENTIAL_DEPOSIT;
```

**Comment by apopiak** [comment]
the statemint <-> statemine typos happen all over the place, should we revisit the naming? xD

**Pull request merged**